docs/code/IDataViewDesignPrinciples.md

```diff
@@ -0,0 +1,479 @@
+# IDataView Design Principles
+
+## Overview
+
+### Brief Introduction to IDataView
+
+The *IDataView system* is a set of interfaces and components that provide
+efficient, compositional processing of schematized data for machine learning
+and advanced analytics applications. It is designed to gracefully and
+efficiently handle high dimensional data and large data sets. It does not
+directly address distributed data and computation, but is suitable for single
+node processing of data partitions belonging to larger distributed data sets.
+
+IDataView is the data pipeline machinery for ML.NET. Microsoft teams consuming
+this library have implemented libraries of IDataView related components
+(loaders, transforms, savers, trainers, predictors, etc.) and has validated
+the performance, scalability and task flexibility benefits.
+
+The name IDataView was inspired from the database world, where the term table
+typically indicates a mutable body of data, while a view is the result of a
+query on one or more tables or views, and is generally immutable. Note that
+both tables and views are schematized, being organized into typed columns and
+rows conforming to the column types. Views differ from tables in several ways:
+
+* Views are *composable*. New views are formed by applying transformations
+  (queries) to other views. In contrast, forming a new table from an existing
+  table involves  copying data, making the tables decoupled; the new table is
+  not linked to the original table in any way.
+
+* Views are *virtual*; tables are fully realized/persisted. In other words, a
+  table contains the values in the rows while a view computes values from
+  other views or  tables, so does not contain or own the values.
+
+* Views are *immutable*; tables are mutable. Since a view does not contain
+  values, but merely computes values from its source views, there is no
+  mechanism for modifying the values.
+
+Note that immutability and compositionality are critical enablers of
+technologies that require reasoning over transformation, like query
+optimization and remoting. Immutability is also key for concurrency and thread
+safety. Views being virtual minimizes I/O, memory allocation, and
+computation—information is accessed, memory is allocated, and computation is
+performed, only when needed to satisfy a local request for information.
+
+### Design Requirements
+
+The IDataView design fulfills the following design requirements:
+
+* **General schema**: Each view carries schema information, which specifies
+  the names and types of the view's columns, together with metadata associated
+  with the  columns. The system is optimized for a reasonably small number of
+  columns (hundreds). See [here](#basics).
+
+* **Open type system**: The column type system is open, in the sense that new
+  data types can be introduced at any time and in any assembly. There is a set
+  of  standard types (which may grow over time), but there is no registry of
```

Comment:
>   [](start = 4, length = 2)

did we do some search and replace that caused these? won't comment on these we should just fix all.

Reply:
I'm not sure. Possibly a word-wrap gone wrong? It appears to only have happened in the lists, so maybe when I'm using rewrap it somehow got confused. Anyway I think I was able to catch them.

---
In reply to: [188820989](https://github.com/dotnet/machinelearning/pull/173#discussion_r188820989) [](ancestors = 188820989)